blockchain/chainquery.go

```diff
@@ -156,83 +156,81 @@ func (b *BlockChain) BestInvalidHeader() chainhash.Hash {
 	return hash
 }
 
-// NextNeededBlocks returns hashes for the next blocks after the current best
-// chain tip that are needed to make progress towards the current best known
-// header skipping any blocks that are already known or in the provided map of
-// blocks to exclude.  Typically the caller would want to exclude all blocks
-// that have outstanding requests.
+// PutNextNeededBlocks populates the provided slice with hashes for the next
+// blocks after the current best chain tip that are needed to make progress
+// towards the current best known header skipping any blocks that already have
+// their data available.
 //
-// The maximum number of results is limited to the provided value or the maximum
-// allowed by the internal lookahead buffer in the case the requested number of
-// max results exceeds that value.
+// The provided slice will be populated with either as many hashes as it will
+// fit per its length or as many hashes it takes to reach best header, whichever
+// is smaller.
+//
+// It returns the a sub slice of the provided one with its bounds adjusted to
+// the number of entries populated.
 //
 // This function is safe for concurrent access.
-func (b *BlockChain) NextNeededBlocks(maxResults uint8, exclude map[chainhash.Hash]struct{}) []*chainhash.Hash {
+func (b *BlockChain) PutNextNeededBlocks(out []chainhash.Hash) []chainhash.Hash {
+	b.index.RLock()
+	defer b.index.RUnlock()
```

Comment:
It will pretty much never be called with zero in practice.  I moved the lock as a defensive measure because the parameter is being mutated, but the caller really needs to protect it anyway if it intends to use the result concurrently, so it doesn't really matter.  I went ahead and moved it after.